### PR TITLE
Add IShaderCompiler api to bob for custom compilation of shaders

### PR DESCRIPTION
**Pull request by JCash**
## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

### CODE REVIEW

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/FragmentProgramBuilder.java [hunk 1]
```diff
@@ -19,15 +19,40 @@
...
+
+            String platformKey = project.getPlatformStrings()[0];
+            Platform platform = Platform.get(platformKey);
+            IShaderCompiler shaderCompiler = project.getShaderCompiler(platform);
+            ShaderDesc shaderDesc = shaderCompiler.compile(is, SHADER_TYPE, in, task.getOutputs().get(0).getPath(), isDebug, outputSpirv, soft_fail);
+            task.output(0).setContent(shaderDesc.toByteArray());
+        }
+    }
+    */
+
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/FragmentProgramBuilder.java:55]: Junk!

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/VertexProgramBuilder.java [hunk 2]
```diff
@@ -19,15 +19,43 @@
...
+            Platform platform = Platform.get(platformKey);
+
+            IShaderCompiler shaderCompiler = project.getShaderCompiler(platform);
+            ShaderDesc shaderDesc = shaderCompiler.compile(is, SHADER_TYPE, in, task.getOutputs().get(0).getPath(), isDebug, outputSpirv, soft_fail);
+
+            task.output(0).setContent(shaderDesc.toByteArray());
+        }
+    }
+    */
+
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/VertexProgramBuilder.java:58]: Junk!

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/VertexProgramBuilder.java [hunk 3]
```diff
@@ -19,15 +19,43 @@
...
+import com.dynamo.bob.Project;
+import com.dynamo.bob.fs.DefaultFileSystem;
+import com.dynamo.bob.fs.IResource;
 import com.dynamo.bob.pipeline.ShaderUtil.ES2ToES3Converter;
+import com.dynamo.bob.pipeline.IShaderCompiler;
+import com.dynamo.bob.pipeline.ShaderCompilers;
+import com.dynamo.bob.pipeline.ShaderPreprocessor;
+import com.dynamo.graphics.proto.Graphics.ShaderDesc;
+
+import org.apache.commons.cli.CommandLine;
```
Comment (Jhonnyg) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/VertexProgramBuilder.java:32]: I wonder if we actually use all these now that we have simplified the builder

File: share/ext/spirv-tools/build_spirv-tools.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env bash
+# License: MIT
+# Copyright 2022 The Defold Foundation
```
Comment (Jhonnyg) [inline share/ext/spirv-tools/build_spirv-tools.sh:3]: I guess this should say 2023, or just omit the year

File: share/ext/spirv-tools/build_spirv-tools.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+EXE_SUFFIX=
+case $PLATFORM in
+    win32|x86_64-win32)
+        EXE_SUFFIX=.exe
+        SRC_EXE=./tools/Release/spirv-opt${EXE_SUFFIX}
+        ;;
+    *)
+        SRC_EXE=./tools/spirv-opt${EXE_SUFFIX}
+        ;;
+esac
```
Comment (Jhonnyg) [inline share/ext/spirv-tools/build_spirv-tools.sh:51]: For future reference maybe we should add a comment that this *only* packages the spirv-opt tool. I think there might be compile options for which items to build, but I didn't spend that much time optimizing this setup..

**Review by Jhonnyg — DISMISSED** [review]
Nice! There's some minor cleanup but otherwise it looks good to go :)

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/FragmentProgramBuilder.java (hunk 1 above)
Comment (JCash) [reply com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/FragmentProgramBuilder.java:55]: I thought you left it there for later :)
I'll remove it then.

**Pull request merged**